gateway/handler_unixfs_dir.go

```diff
@@ -173,11 +177,9 @@ func (i *handler) serveDirectory(ctx context.Context, w http.ResponseWriter, r *
 		}
 	}
 
-	size := "?"
-	if s, err := dir.Size(); err == nil {
-		// Size may not be defined/supported. Continue anyways.
-		size = humanize.Bytes(uint64(s))
-	}
+	// TODO: Why was size allowed to be undefined/supported with continuing anyways? Is this an oversight?
```

Comment:
Because the function `.Size()` is allowed to not be implemented, therefore it can error:

https://github.com/ipfs/go-libipfs/blob/124ae4b3a578bd8e53f16b4e6d460040504d840a/files/file.go#L20-L23

Reply:
But these were always UnixFS file and implemented right? I'm fine if we want to keep allowing this behavior, but only if we have a use for it. Otherwise we require sizes and make a change later if we need it.

Reply:
@aschmahmann I found the source of that change: https://github.com/ipfs/kubo/commit/044790a83857c4b396ce5c336045d6b10f5e4075 - perhaps your past self can tell us the reason 😅 Posibly just something that got overlooked?

Reply:
> perhaps your past self can tell us the reason 😅

🤣, but unfortunately you've got to do a bit more GitHub archaeology since it's even further back. It looks like it was actually implemented in https://github.com/ipfs/kubo/pull/6680. Looks like there was a bug in rendering directories that had symlinks in them due to symlinks not having sizes and the solution was to fix dealing with symlinks and to protect against unknown sizes causing problems in the future (AFAICT a defensive measure).

I'm fine putting the defensive measure back in/making the size nilable so implementations don't have to pass it if they don't want it/have it and they just get a `?` in the directory index.html for the sizes. Don't feel strongly either way, might be worth keeping just in case (makes the bugs people file look like "hey I got a ? for size" rather than "hey I got a 500 error").

Reply:
Ahah, yes, I think making it nillable is a better idea, so that we can keep that safe-guard in case the size isn't known.

Reply:
@aschmahmann maybe, an interesting idea would be to totally remove the size requirement for the directory and simply sum the cumulative size (from the links). Our goal is to give an estimation, and it may never be byte-correct since the values in the links can always be faked. This could further simplify the type `directoryResponse`. Wdyt?

Reply:
I believe the origin of `?` was to not crash directory listing because size is invalid or missing.  There may be malformed UnixFS DAGs which should still be enumerable and user should still be able to traverse it by clicking on its children. 

Gateway's purpose is to enable access to content-addressed data, so we have to be liberal with inputs. If we want to fix UnixFS, better to improve libraries, and specs.

With that in mind:
- at least don't crash if size info is missing, enumerate children
- ideally, we should try to use total DAG size if present, and if that is missing or invalid (<0), do the estimate by adding up all Tsizes from links, so user always gets something useful. we already have disclaimers in dir-index-html that the size reflects entire sub-DAG, and not the data alone.